src/java/htsjdk/samtools/SAMRecord.java

```diff
@@ -1832,6 +1832,87 @@ public Object clone() throws CloneNotSupportedException {
+     *
+     */
+    public SAMRecord deepCopy() {
+        final SAMRecord newSAM = new SAMRecord(getHeader());
+
+        newSAM.setReadName(getReadName());
+        newSAM.setReadBases(Arrays.copyOf(getReadBases(), getReadLength()));
+        final byte baseQualities[] = getBaseQualities();
+        newSAM.setBaseQualities(Arrays.copyOf(baseQualities, baseQualities.length));
+        newSAM.setReferenceName(getReferenceName());
+        newSAM.setAlignmentStart(getAlignmentStart()); // sets mAlignmentEnd
+        newSAM.setMappingQuality(getMappingQuality());
+        newSAM.setCigarString(getCigarString()); // clears Cigar element and alignmentBlocks
+        newSAM.setFileSource(null);
+
+        newSAM.setFlags(getFlags());
+        newSAM.setMateReferenceName(getMateReferenceName());
+        newSAM.setMateAlignmentStart(getMateAlignmentStart());
+        newSAM.setInferredInsertSize(getInferredInsertSize());
+        newSAM.setReferenceIndex(getReferenceIndex());
+        newSAM.setMateReferenceName(getMateReferenceName());
+        newSAM.setMateReferenceIndex(getMateReferenceIndex());
+        newSAM.setValidationStringency(getValidationStringency());
+
+        SAMBinaryTagAndValue attributes = getBinaryAttributes();
+        while (null != attributes) {
+                final Object value = attributes.value;
+                final String tag = SAMTagUtil.getSingleton().makeStringTag(attributes.tag);
+                if (value instanceof byte[]) {
+                    if (isUnsignedArrayAttribute(tag)) {
+                        newSAM.setUnsignedArrayAttribute(tag, ((byte[]) value).clone());
+                    }
+                    else {
+                        newSAM.setAttribute(tag, ((byte[]) value).clone());
+                    }
+                }
+                else if (value instanceof short[]) {
+                    if (isUnsignedArrayAttribute(tag)) {
+                        newSAM.setUnsignedArrayAttribute(tag, ((short[]) value).clone());
+                    }
+                    else {
+                        newSAM.setAttribute(tag, ((short[]) value).clone());
+                    }
+                }
+                else if (value instanceof int[]) {
+                    if (isUnsignedArrayAttribute(tag)) {
+                        newSAM.setUnsignedArrayAttribute(tag, ((int[]) value).clone());
+                    }
+                    else {
+                        newSAM.setAttribute(tag, ((int[]) value).clone());
+                    }
+                }
+                else if (value instanceof float[]) {
+                    newSAM.setAttribute(tag, ((float[]) value).clone());
+                }
+                else {
+                    // otherwise, the api limits the remaining possible value types to
+                    // immutable (String or boxed primitive) types
+                    newSAM.setAttribute(tag, value);
+                }
+            attributes = attributes.getNext();
```

Comment:
Just noticed that `SAMBinaryTagAndValue` already has a `copy()` method that claims to perform a deep copy, but doesn't. I recommend that you update the docs for `SAMBinaryTagAndValue.copy()` to state that it performs a shallow copy, move the code above into a new `SAMBinaryTagAndValue.deepCopy()` method, and call that from `SAMRecord.deepCopy()`